File: src/test/java/com/github/squirrelgrip/meter/service/Line900ProcessorTest.java
package com.github.squirrelgrip.meter.service;

import com.github.squirrelgrip.meter.database.DatabaseSession;
import com.github.squirrelgrip.meter.domain.ContextKey;
import com.github.squirrelgrip.meter.exception.VersionNotSetException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.HashMap;

import static com.github.squirrelgrip.meter.domain.ContextKey.*;
import static com.github.squirrelgrip.meter.domain.Version.NEM12;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class Line900ProcessorTest {
    private Line900Processor testSubject = new Line900Processor();
    private HashMap<ContextKey, Object> context;
    @Mock
    private DatabaseSession databaseSession;

    @BeforeEach
    public void beforeEach() {
        context = new HashMap<>();
        context.put(LINE_NUMBER, 2);
        context.put(VERSION, NEM12);
        context.put(SESSION, databaseSession);
    }

    @Test
    public void process_Given900RecordAndVersionNotSet() {
        context.remove(VERSION);

        VersionNotSetException exception = assertThrows(VersionNotSetException.class, () -> testSubject.preProcess(context));
        assertThat(exception.getMessage()).isEqualTo("Version not set");
    }

    @Test
    public void preProcess() {
        context.put(NMI, "abcd");
        context.put(INTERVAL_LENGTH, 5);

        testSubject.preProcess(context);

        assertThat(context).doesNotContainKey(NMI);
        assertThat(context).doesNotContainKey(INTERVAL_LENGTH);
    }

    @Test
    public void process() {
        testSubject.process("900".split(","), context);

        assertThat(context).doesNotContainKey(VERSION);
        verify(databaseSession).commit();
    }

}

File: src/test/kotlin/com/github/squirrelgrip/meter/service/Line900ProcessorTest.kt
package com.github.squirrelgrip.meter.service

import com.github.squirrelgrip.meter.database.DatabaseSession
import com.github.squirrelgrip.meter.domain.BaseMeterContext
import com.github.squirrelgrip.meter.domain.ContextKey
import com.github.squirrelgrip.meter.domain.MeterContext
import com.github.squirrelgrip.meter.domain.Version
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
internal class Line900ProcessorTest {
    private lateinit var testSubject: Line900Processor
    private lateinit var context: MeterContext

    @Mock
    private lateinit var session: DatabaseSession

    @BeforeEach
    fun beforeEach() {
        testSubject = Line900Processor()
        context = BaseMeterContext(session, "")
        context.put(ContextKey.LINE_NUMBER, 2)
        context.put(ContextKey.VERSION, Version.NEM12)
        context.put(ContextKey.SESSION, session)
    }

    @Test
    fun process_Given() {
        context.put(ContextKey.NMI, "abcd")
        context.put(ContextKey.INTERVAL_LENGTH, 5)

        testSubject.process(context, 1, "500,")

        assertThat(context.currentMap).doesNotContainKey(ContextKey.NMI)
        assertThat(context.currentMap).doesNotContainKey(ContextKey.INTERVAL_LENGTH)
    }

    @Test
    fun process() {
        testSubject.process(context, 1, "900")

        assertThat(context.currentMap).doesNotContainKey(ContextKey.VERSION)

        Mockito.verify(session).commit()
    }
}